stdpopsim/slim_engine.py

```diff
@@ -0,0 +1,663 @@
+            g = G_start + gdiff(T_0, subpopulation_splits[0,i]);
+            newpop = subpopulation_splits[1,i];
+            size = asInteger(subpopulation_splits[2,i] / Q);
+            oldpop = subpopulation_splits[3,i];
+            sim.registerLateEvent(NULL,
+                "{dbg(self.source); " +
+                "sim.addSubpopSplit("+newpop+","+size+","+oldpop+");}",
+                g, g);
+        }
+    }
+
+    // Population size changes.
+    if (num_epochs > 1) {
+        for (i in 1:(num_epochs-1)) {
+            g = G[i-1];
+            for (j in 0:(num_populations-1)) {
+                if (N[i,j] != N[i-1,j]) {
+                    sim.registerLateEvent(NULL,
+                        "{dbg(self.source); " +
+                        "p"+j+".setSubpopulationSize("+N[i,j]+");}",
+                        g, g);
+                }
+
+                if (growth_rates[i,j] != 0) {
+                    growth_phase_start = g+1;
+                    if (i == num_epochs-1)
+                        growth_phase_end = G[i];
+                    else {
+                        // We already registered a size change at generation G[i].
+                        growth_phase_end = G[i] - 1;
+                    }
+
+                    if (growth_phase_start >= growth_phase_end) {
+                        // Some demographic models have duplicate epoch times,
+                        // which should be ignored.
+                        next;
+                    }
+
+                    N0 = N[i,j];
+                    r = Q * growth_rates[i,j];
+
+                    sim.registerLateEvent(NULL,
+                        "{" +
+                            "dbg(self.source); " +
+                            "gx=sim.generation-"+g+"; " +
+                            "size=asInteger(round("+N0+"*exp("+r+"*gx))); " +
+                            "p"+j+".setSubpopulationSize(size);" +
+                        "}",
+                        growth_phase_start, growth_phase_end);
+                }
+            }
+        }
+
+        // Migration rates.
+        for (i in 1:(num_epochs-1)) {
+            for (j in 0:(num_populations-1)) {
+                for (k in 0:(num_populations-1)) {
+                    if (j==k | N[i,j] == 0 | N[i,k] == 0)
+                        next;
+
+                    m_last = migration_matrices[k,j,i-1];
+                    m = migration_matrices[k,j,i];
+                    if (m == m_last) {
+                        // Do nothing if the migration rate hasn't changed.
+                        next;
+                    }
+                    g = G[i-1];
+                    sim.registerLateEvent(NULL,
+                        "{dbg(self.source); " +
+                        "p"+j+".setMigrationRates("+k+", "+m+");}",
+                        g, g);
+                }
+            }
+        }
+    }
+
+    // Admixture pulses.
+    if (length(admixture_pulses) > 0 ) {
```

Comment:
not sure that there _is_ reason for concern, but does the ordering of these event types (splits, then size changes, then admixture events) bias things in the simulation? are there edge cases we might worry about?

Reply:
I don't think the ordering currently matters. This just sets callbacks, which run in the order in which the callbacks are set up. But the effect of the callbacks are determined by SLiM's generation cycle, and this is not affected by the order of callbacks here.

The order of callbacks can and does matter in more complex cases involving saving/reloading simulation state (used with selective sweeps). We can certainly discuss event ordering again down the road when I submit a PR with selection bits.

Reply:
Oh, there are some limitations on event ordering here. Subpopulation splits create new populations, and the new population must exist before they can be manipulated in other events (if any). But no bias here, I don't think.